sparkles/roll_optimize.py

```diff
@@ -207,9 +214,47 @@ def get_ids_list(roll_offsets):
         ids_list = get_ids_list(roll_offsets)
         ids0 = ids_list[len(ro_minus)]
 
+        get_roll_intervals_func = getattr(self, f'_get_roll_intervals_{method}')
+        roll_intervals = get_roll_intervals_func(ids0, ids_list, roll_targ, roll_min,
+                                                 roll_max, roll_offsets, d_roll)
+
+        roll_info = {'roll_min': roll_min,
+                     'roll_max': roll_max,
+                     'roll_nom': roll_nom}
+
+        return sorted(roll_intervals, key=lambda x: x['roll']), roll_info
+
+    @staticmethod
+    def _get_roll_intervals_uniform(ids0, ids_list, roll_targ, roll_min, roll_max,
+                                    roll_offsets, d_roll):
+        """Private method to get uniform set of roll intervals over allowed range.
+        """
+        roll_intervals = []
+        for roll_offset, ids in zip(roll_offsets, ids_list):
+            roll_rolled = roll_targ + roll_offset
+            if roll_rolled < roll_min or roll_rolled > roll_max:
+                continue
+
+            roll_interval = {'roll': roll_rolled,
+                             'roll_min': roll_rolled - d_roll / 2,
+                             'roll_max': roll_rolled + d_roll / 2,
+                             'add_ids': ids - ids0,
+                             'drop_ids': ids0 - ids}
+
+            roll_intervals.append(roll_interval)
+
+        return roll_intervals
+
+    @staticmethod
+    def _get_roll_intervals_uniq_ids(ids0, ids_list, roll, roll_min, roll_max,
+                                     roll_offsets, d_roll):
+        """Private method to get roll intervals that span a range where there is a unique
+        set of available candidate stars within the entire interval.
+
+        """
         # Get all unique sets of stars that are in the FOV over the sampled
         # roll offsets.  Ignore ids sets that do not add new candidate stars.
-        uniq_ids_sets = []
+        uniq_ids_sets = [ids0]
```

Comment:
@jeanconn - this is the one line of code that implements "it is also improves the original algorithm by including the roll interval corresponding to the original MP-requested roll, but using a roll at the center of the roll interval." It will probably be much more obvious here in the code, I did struggle trying to describe it and knew that what I wrote was not really adequate.

Basically it generates a list of roll intervals and in the previous code version it was excluding the roll interval which includes the original roll angle. This was a problem for catalogs where there is an acq star very near the FOV edge which is excluded as guide. In this case it would not allow rolling that star into the corner where it would be OK for both acq and guide.